<a href="https://colab.research.google.com/github/gumdropsteve/intro_to_machine_learning/blob/main/day_10/04_regression_weekend_project_solution.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Regressed Classification Weekend Project

### **Challenge Options**

#### Saudi Airbnb Price Prediction
- Data: [github.com/gumdropsteve/intro_to_machine_learning/tree/main/day_10/data](https://github.com/gumdropsteve/intro_to_machine_learning/tree/main/day_10/data)
- Data Dictionary: [github.com/gumdropsteve/intro_to_machine_learning/tree/main/day_10/data/README.md](https://github.com/gumdropsteve/intro_to_machine_learning/tree/main/day_10/data/README.md)
- Type: regression
- Starter code: [gist.github.com/gumdropsteve/c178891652838cb0663a4ef1d2eb9d14](https://gist.github.com/gumdropsteve/c178891652838cb0663a4ef1d2eb9d14)
- Target: `price`

#### Income Bracket Prediction
- Data: [archive.ics.uci.edu/ml/machine-learning-databases/adult/](https://archive.ics.uci.edu/ml/machine-learning-databases/adult/)
- Data Dictionary: [archive.ics.uci.edu/ml/datasets/Adult](https://archive.ics.uci.edu/ml/datasets/Adult)
- Type: classification
- Starter code: [gist.github.com/gumdropsteve/25e40b0ab0b5a6a7d51f11c00f91d0bb](https://gist.github.com/gumdropsteve/25e40b0ab0b5a6a7d51f11c00f91d0bb)
- Target: `target`

#### Mushroom
- Data: [archive.ics.uci.edu/ml/machine-learning-databases/mushroom/](https://archive.ics.uci.edu/ml/machine-learning-databases/mushroom/)
- Data Dictionary: [archive.ics.uci.edu/ml/datasets/mushroom](https://archive.ics.uci.edu/ml/datasets/mushroom)
- Type: classification
- Starter code: [colab.research.google.com/drive/1hIoS2jZCpVOCE4U2mxqkeYPLh2m6I8rA?usp=sharing](https://colab.research.google.com/drive/1hIoS2jZCpVOCE4U2mxqkeYPLh2m6I8rA?usp=sharing)
- Target: `target`

### **Available Models**
Pick from these models based on which challenge you take on (regression or classification). 

#### Regression
- [LinearRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html)
- [SGDRegressor](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDRegressor.html)
- [KNeighborsRegressor](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsRegressor.html)

#### Classification
- [LogisticRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html)
- [SGDClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html)
- [KNeighborsClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html)

## **Deleverables**
1. EDA
   - What is this data?
   - Visualize some relationships
   - What do you see? What did you find?
2. Data prep & `train_test_split()`
    - Make this data ready for a ML model
    - Decide which features to include
    - Split into training and testing datasets
3. Baseline model
    - What is your baseline model?
        - Explain what's going on please... Assume I'm an idiot 
4. Create a model
    - Compare your results to the baseline
5. Optimize your model
    - [GridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html)
        - Bonus: [RandomizedSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html)
    - Compare your results to the baseline
        - How did you do this? Why?
    - If you don't beat the baseline, try improving your model or using a different model
        - Were you able to beat the baseline? Why?

In [ ]:
import pandas as pd

dammam_link = 'https://github.com/gumdropsteve/intro_to_machine_learning/raw/main/day_10/data/dammam.parquet'
jeddah_link = 'https://github.com/gumdropsteve/intro_to_machine_learning/raw/main/day_10/data/jeddah.parquet'
riyadh_link = 'https://github.com/gumdropsteve/intro_to_machine_learning/raw/main/day_10/data/riyadh.parquet'

df = pd.read_parquet(jeddah_link)

df

ds  ... chefs_kitchen_bool
0      2021-01-20 04:13:49.541603  ...              False
1      2021-01-20 04:13:49.541603  ...              False
2      2021-01-20 04:13:49.541603  ...              False
3      2021-01-20 04:13:49.541603  ...              False
4      2021-01-20 04:13:49.541603  ...              False
...                           ...  ...                ...
10115  2021-02-01 05:23:57.652792  ...              False
10116  2021-02-01 05:23:57.652792  ...              False
10117  2021-02-01 05:23:57.652792  ...              False
10118  2021-02-01 05:23:57.652792  ...              False
10119  2021-02-01 05:23:57.652792  ...              False

[10120 rows x 40 columns]

## Baseline Model

In [ ]:
def baseline_model(n_predictions, value_to_predict):
    """
    just predict a single value (e.g. mean) for everything
    """
    baseline_preds = []
    for i in range(n_predictions):
        baseline_preds.append(value_to_predict)
    return pd.Series(baseline_preds)

In [ ]:
from sklearn.model_selection import train_test_split

X = df
y = df.price

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, random_state=4)

In [ ]:
import numpy as np

n_predictions = len(y_test)  # how many predictions to make?
baseline_value = np.mean(y_train)  # what value to predict?

baseline_preds = baseline_model(n_predictions, baseline_value)

What are we predicting? Note: all predictions should be the same because this is a simple baseline model..

In [ ]:
baseline_preds.tail(5)

3031    293.866742
3032    293.866742
3033    293.866742
3034    293.866742
3035    293.866742
dtype: float64

In [ ]:
from sklearn.metrics import mean_absolute_error

mean_absolute_error(y_test, baseline_preds)

366.39182603785633

In [ ]:
from sklearn.metrics import mean_squared_error

mean_squared_error(y_test, baseline_preds)

1574818.8716798488

## Simple Solve
Getting a MVP out the door...

In [ ]:
df.columns

Index(['ds', 'search_filter', 'url', 'title', 'type', 'location', 'guests',
       'bedrooms', 'beds', 'is_studio', 'baths', 'half_baths', 'shared_baths',
       'price', 'avg_rating', 'n_reviews', 'gym_bool', 'wifi_bool',
       'self_check_in_bool', 'air_conditioning_bool', 'pets_allowed_bool',
       'indoor_fireplace_bool', 'hot_tub_bool', 'free_parking_bool',
       'pool_bool', 'kitchen_bool', 'breakfast_bool', 'elevator_bool',
       'washer_bool', 'dryer_bool', 'heating_bool', 'waterfront_bool',
       'dishwasher_bool', 'beachfront_bool', 'ski_in_ski_out_bool',
       'terrace_bool', 'sonos_sound_system_bool', 'bbq_grill_bool',
       'hair_dryer_bool', 'chefs_kitchen_bool'],
      dtype='object')

In [ ]:
corr = df.corr()

corr = corr.head(10)

corr.style.background_gradient(cmap='coolwarm')

/usr/local/lib/python3.6/dist-packages/pandas/io/formats/style.py:1126: RuntimeWarning: All-NaN slice encountered
  smin = np.nanmin(s.to_numpy()) if vmin is None else vmin
/usr/local/lib/python3.6/dist-packages/pandas/io/formats/style.py:1127: RuntimeWarning: All-NaN slice encountered
  smax = np.nanmax(s.to_numpy()) if vmax is None else vmax


In [ ]:
# display values from row 7 (price)
corr.iloc[7]

guests                     0.113631
bedrooms                   0.087197
beds                       0.113631
is_studio                  0.023333
baths                      0.101232
half_baths                      NaN
shared_baths              -0.035664
price                      1.000000
avg_rating                 0.153463
n_reviews                 -0.102559
gym_bool                  -0.004810
wifi_bool                  0.098935
self_check_in_bool        -0.078175
air_conditioning_bool      0.026414
pets_allowed_bool         -0.011561
indoor_fireplace_bool     -0.030433
hot_tub_bool              -0.019329
free_parking_bool         -0.073309
pool_bool                  0.217863
kitchen_bool              -0.006960
breakfast_bool            -0.011980
elevator_bool             -0.021975
washer_bool                0.025910
dryer_bool                 0.069911
heating_bool                    NaN
waterfront_bool            0.013711
dishwasher_bool           -0.005414
beachfront_bool            0

In [ ]:
import numpy as np

np.sum(df.isnull())

ds                             0
search_filter               3300
url                            0
title                          0
type                           0
location                       0
guests                      4002
bedrooms                       0
beds                        4002
is_studio                      0
baths                       1535
half_baths                 10056
shared_baths                9911
price                          0
avg_rating                  5193
n_reviews                   5193
gym_bool                       0
wifi_bool                      0
self_check_in_bool             0
air_conditioning_bool          0
pets_allowed_bool              0
indoor_fireplace_bool          0
hot_tub_bool                   0
free_parking_bool              0
pool_bool                      0
kitchen_bool                   0
breakfast_bool                 0
elevator_bool                  0
washer_bool                    0
dryer_bool                     0
heating_bo

In [ ]:
from sklearn.model_selection import train_test_split

X = df[['bedrooms', 'baths', 'shared_baths', 'avg_rating', 'n_reviews', 
        'gym_bool', 'pool_bool', 'dishwasher_bool']]
y = df.price

# all missing values here likely mean None, so let's fill with 0
X = X.fillna(0)

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, random_state=4)

In [ ]:
len(X_train), len(y_test)

(7084, 3036)

In [ ]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()

lr.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
preds = lr.predict(X_test)

# show the last 5 predictions
preds[-5: ]

array([330.0269055 , 351.4583746 , 430.79635835, 380.41163192,
       122.78438572])

In [ ]:
from sklearn.metrics import r2_score

r2_score(y_test, preds)

0.06680379957991911

In [ ]:
from sklearn.metrics import mean_absolute_error

mean_absolute_error(y_test, preds)

368.6388682911098

In [ ]:
from sklearn.metrics import mean_squared_error

mean_squared_error(y_test, preds)

1469599.093084856

#### Adding a quick categorical feature to that...
Turning `type` into numbers because I have a feeling it might help..

In [ ]:
df.type.unique()

array(['Entire loft', 'Entire apartment', 'Entire condominium',
       'Entire serviced apartment', 'Entire house', 'Private room',
       'Entire villa', 'Entire guest suite', 'Shared room',
       'Entire townhouse', 'Tiny house', 'Room', 'Entire chalet',
       'Hotel room', 'Resort room', 'Entire place', 'Entire cottage',
       'Earth house', 'Entire bungalow', 'Farm stay', 'Cycladic house'],
      dtype=object)

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.get_dummies.html

In [ ]:
pd.get_dummies(df.type)

Cycladic house  Earth house  ...  Shared room  Tiny house
0                   0            0  ...            0           0
1                   0            0  ...            0           0
2                   0            0  ...            0           0
3                   0            0  ...            0           0
4                   0            0  ...            0           0
...               ...          ...  ...          ...         ...
10115               0            0  ...            0           0
10116               0            0  ...            0           0
10117               0            0  ...            0           0
10118               0            0  ...            0           0
10119               0            0  ...            0           0

[10120 rows x 21 columns]

In [ ]:
X = df[['bedrooms', 'baths', 'shared_baths', 'avg_rating', 'n_reviews', 'gym_bool', 'pool_bool', 'dishwasher_bool']]

pd.concat([X, pd.get_dummies(df['type'])], axis=1)

bedrooms  baths  shared_baths  ...  Room  Shared room  Tiny house
0           1.0    1.5           NaN  ...     0            0           0
1           1.0    1.5           NaN  ...     0            0           0
2           2.0    2.0           NaN  ...     0            0           0
3           2.0    1.0           NaN  ...     0            0           0
4           1.0    1.0           NaN  ...     0            0           0
...         ...    ...           ...  ...   ...          ...         ...
10115       2.0    2.0           NaN  ...     0            0           0
10116       2.0    1.0           NaN  ...     0            0           0
10117       2.0    2.0           NaN  ...     0            0           0
10118       1.0    1.0           NaN  ...     0            0           0
10119       2.0    5.5           NaN  ...     0            0           0

[10120 rows x 29 columns]

In [ ]:
X = pd.concat([X, pd.get_dummies(df['type'])], axis=1)
y = df.price

# fill missing values in X with 0
X = X.fillna(0)

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, random_state=4)

In [ ]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()

lr.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
preds = lr.predict(X_test)

r2_score(y_test, preds)

0.27028494682784765

In [ ]:
from sklearn.metrics import mean_absolute_error

mean_absolute_error(y_test, preds)

439.6503004282715

In [ ]:
from sklearn.metrics import mean_squared_error

mean_squared_error(y_test, preds)

1149156.6080845848

It helped a little! :)

## Avoiding Leakage
Improving the quality of the data...

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10120 entries, 0 to 10119
Data columns (total 40 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ds                       10120 non-null  object 
 1   search_filter            6820 non-null   object 
 2   url                      10120 non-null  object 
 3   title                    10120 non-null  object 
 4   type                     10120 non-null  object 
 5   location                 10120 non-null  object 
 6   guests                   6118 non-null   float64
 7   bedrooms                 10120 non-null  float64
 8   beds                     6118 non-null   float64
 9   is_studio                10120 non-null  bool   
 10  baths                    8585 non-null   float64
 11  half_baths               64 non-null     float64
 12  shared_baths             209 non-null    float64
 13  price                    10120 non-null  float64
 14  avg_rating            

In [ ]:
df.describe()

guests      bedrooms  ...   avg_rating    n_reviews
count  6118.000000  10120.000000  ...  4927.000000  4927.000000
mean      2.603465      1.488834  ...     4.719351    19.680536
std       1.788836      1.043132  ...     0.326824    23.585304
min       0.000000      0.000000  ...     3.330000     3.000000
25%       2.000000      1.000000  ...     4.640000     5.000000
50%       2.000000      1.000000  ...     4.790000    11.000000
75%       3.000000      2.000000  ...     4.960000    25.000000
max      19.000000     10.000000  ...     5.000000   154.000000

[8 rows x 9 columns]

`url` can be used to identify a listing. As no 2 listings could share the same URL, each unique value in `url` represents 1 unique listing.. 

Source: [Data dictionary](https://github.com/gumdropsteve/intro_to_machine_learning/blob/main/day_10/data/README.md#data-dictionary)

In [ ]:
len(df['url'].unique())

342

Because we have n (say 342) unique values in `url`, we have n (say 342) unique listings...

To avoid data leakage, we need to only keep 1 instance of each listing (`url`)... So let's drop the `ds` (datetime the info was pulled) and `search_filter` (search filter applied on top of the city filter) because the same listing (`url`) could show up multiple times under different `ds` and `search_filter` values...

Method: [pd.DataFrame.drop_duplicates() docs](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html)

In [ ]:
len(df.drop(['ds', 'search_filter'], axis=1).drop_duplicates())

1874

That didn't cut it down to all the way... But if we add `subset='url'` then it should...

In [ ]:
len(df.drop(['ds', 'search_filter'], axis=1).drop_duplicates(subset='url'))

342

By adding `subset='url'` we focus specifically on duplicate values in the `url` column..

In [ ]:
dropped_df = df.drop(['ds', 'search_filter'], axis=1).drop_duplicates(subset='url')

dropped_df.sample(3)

url  ... chefs_kitchen_bool
56    http://airbnb.com/rooms/47496281  ...              False
454   http://airbnb.com/rooms/42166199  ...              False
1212  http://airbnb.com/rooms/22198505  ...              False

[3 rows x 38 columns]

So these should be the same..

In [ ]:
len(dropped_df), len(dropped_df['url'].unique())

(342, 342)

Now let's split the data and model once more..

In [ ]:
X = dropped_df[['bedrooms', 'baths', 'shared_baths', 'avg_rating', 'n_reviews', 'gym_bool', 'pool_bool', 'dishwasher_bool']]
X = pd.concat([X, pd.get_dummies(dropped_df['type'])], axis=1)

y = dropped_df.price

# fill missing values in X with 0
X = X.fillna(0)

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, random_state=4)

In [ ]:
len(X_train), len(y_test)

(239, 103)

In [ ]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()

lr.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
preds = lr.predict(X_test)

# show the last 5 predictions
preds[-5: ]

array([243.01319869, 243.01319869, 196.00430459, 244.5139995 ,
       539.48229072])

In [ ]:
from sklearn.metrics import r2_score

r2_score(y_test, preds)

0.3192499479275578

In [ ]:
from sklearn.metrics import mean_absolute_error

mean_absolute_error(y_test, preds)

585.2371916328502

In [ ]:
from sklearn.metrics import mean_squared_error

mean_squared_error(y_test, preds)

2428947.7568374197

Now that's an honest score!

### What about `location`?



In [ ]:
dropped_df.location.value_counts()

Jeddah                297
جدة                    12
aparthotel              7
Jedday                  6
serviced apartment      6
جدّة‎                   4
Jeddah                  4
boutique hotel          2
Al salamah              1
JEDDAH                  1
:                       1
Jeddah city             1
Name: location, dtype: int64

In [ ]:
np.sum(pd.get_dummies(dropped_df.location))

:                       1
Al salamah              1
JEDDAH                  1
Jeddah                297
Jeddah                  4
Jeddah city             1
Jedday                  6
aparthotel              7
boutique hotel          2
serviced apartment      6
جدة                    12
جدّة‎                   4
dtype: int64

In [ ]:
np.sum(pd.get_dummies(dropped_df.location.str.strip()))

:                       1
Al salamah              1
JEDDAH                  1
Jeddah                301
Jeddah city             1
Jedday                  6
aparthotel              7
boutique hotel          2
serviced apartment      6
جدة                    12
جدّة‎                   4
dtype: int64

In [ ]:
np.sum(pd.get_dummies(dropped_df.location.str.strip().str.lower()))

:                       1
al salamah              1
aparthotel              7
boutique hotel          2
jeddah                302
jeddah city             1
jedday                  6
serviced apartment      6
جدة                    12
جدّة‎                   4
dtype: int64

In [ ]:
X = dropped_df[['bedrooms', 'baths', 'shared_baths', 'avg_rating', 'n_reviews', 'gym_bool', 'pool_bool', 'dishwasher_bool']]
X = pd.concat([X, pd.get_dummies(dropped_df['type'])], axis=1)
X = pd.concat([X, pd.get_dummies(dropped_df.location.str.strip().str.lower())], axis=1)

y = dropped_df.price

# fill missing values in X with 0
X = X.fillna(0)

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, random_state=4)

In [ ]:
X_train.sample(3)

bedrooms  baths  shared_baths  ...  serviced apartment  جدة  جدّة‎
270        3.0    4.0           0.0  ...                   0    0      0
274        2.0    1.0           0.0  ...                   0    0      0
1515       1.0    1.0           0.0  ...                   0    0      1

[3 rows x 39 columns]

In [ ]:
len(X_train), len(y_test)

(239, 103)

In [ ]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()

lr.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
preds = lr.predict(X_test)

# show the last 5 predictions
preds[-5: ]

array([290.28881668, 290.28881668, 294.99373993, 259.98974445,
       547.99043834])

In [ ]:
from sklearn.metrics import r2_score

r2_score(y_test, preds)

0.3920955665038459

In [ ]:
from sklearn.metrics import mean_absolute_error

mean_absolute_error(y_test, preds)

592.4381352698242

In [ ]:
from sklearn.metrics import mean_squared_error

mean_squared_error(y_test, preds)

2169031.211406909

## Trying a Different Model (SGDRegressor)

In [ ]:
from sklearn.linear_model import SGDRegressor

sgdr = SGDRegressor()

sgdr.fit(X_train, y_train)

SGDRegressor(alpha=0.0001, average=False, early_stopping=False, epsilon=0.1,
             eta0=0.01, fit_intercept=True, l1_ratio=0.15,
             learning_rate='invscaling', loss='squared_loss', max_iter=1000,
             n_iter_no_change=5, penalty='l2', power_t=0.25, random_state=None,
             shuffle=True, tol=0.001, validation_fraction=0.1, verbose=0,
             warm_start=False)

In [ ]:
preds = sgdr.predict(X_test)

preds[-5: ]

array([-1.38698302e+10, -1.38698302e+10, -7.49524148e+09, -1.24315167e+10,
       -1.52793479e+10])

In [ ]:
r2_score(y_test, preds)

-118309652268788.53

In [ ]:
mean_absolute_error(y_test, preds)

17800971461.39324

In [ ]:
mean_squared_error(y_test, preds)

4.221343261240194e+20

That didn't go well...

## GridSearchCV
[LinearRegression Params](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html)

In [ ]:
from sklearn.model_selection import GridSearchCV

lr = LinearRegression()

param_grid = {'fit_intercept':[True, False],
              'normalize': [False, True]}


grid = GridSearchCV(lr, param_grid)

grid.fit(X_train, y_train)

grid.best_params_

{'fit_intercept': False, 'normalize': False}

In [ ]:
grid.best_estimator_

LinearRegression(copy_X=True, fit_intercept=False, n_jobs=None, normalize=False)

In [ ]:
lr = grid.best_estimator_

lr.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=False, n_jobs=None, normalize=False)

In [ ]:
preds = lr.predict(X_test)

preds[-5: ]

array([290.28881668, 290.28881668, 457.54263568, 259.98974445,
       547.99043834])

In [ ]:
r2_score(y_test, preds)

0.39373482596382325

In [ ]:
mean_absolute_error(y_test, preds)

589.0188219287975

In [ ]:
mean_squared_error(y_test, preds)

2163182.25763002

[SGDR params](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDRegressor.html)

In [ ]:
from sklearn.model_selection import GridSearchCV

sgdr = SGDRegressor()

param_grid = {'loss':['squared_loss', 'huber', 'epsilon_insensitive', 'squared_epsilon_insensitive'],
              'penalty': ['l1', 'l2', 'elasticnet'],
              'alpha': [0.0001, 0.001, 0.01, 0.1, 0.00001],
              'l1_ratio': [0.15, 0.10, 0.05, 0.30],
              'fit_intercept': [True, False],
              'max_iter': [1000, 2000, 3000, 4000, 5000, 6000, 7000],
              'shuffle': [True, False]}

grid = GridSearchCV(sgdr, param_grid)

grid.fit(X_train, y_train)

grid.best_params_

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_stochastic_gradient.py:1187: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_stochastic_gradient.py:1187: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_stochastic_gradient.py:1187: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_stochastic_gradient.py:1187: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linea

{'alpha': 1e-05,
 'fit_intercept': True,
 'l1_ratio': 0.15,
 'loss': 'epsilon_insensitive',
 'max_iter': 1000,
 'penalty': 'l1',
 'shuffle': False}

In [ ]:
grid.best_estimator_

SGDRegressor(alpha=1e-05, average=False, early_stopping=False, epsilon=0.1,
             eta0=0.01, fit_intercept=True, l1_ratio=0.15,
             learning_rate='invscaling', loss='epsilon_insensitive',
             max_iter=1000, n_iter_no_change=5, penalty='l1', power_t=0.25,
             random_state=None, shuffle=False, tol=0.001,
             validation_fraction=0.1, verbose=0, warm_start=False)

In [ ]:
sgdr = grid.best_estimator_

sgdr.fit(X_train, y_train)

SGDRegressor(alpha=1e-05, average=False, early_stopping=False, epsilon=0.1,
             eta0=0.01, fit_intercept=True, l1_ratio=0.15,
             learning_rate='invscaling', loss='epsilon_insensitive',
             max_iter=1000, n_iter_no_change=5, penalty='l1', power_t=0.25,
             random_state=None, shuffle=False, tol=0.001,
             validation_fraction=0.1, verbose=0, warm_start=False)

In [ ]:
preds = sgdr.predict(X_test)

preds[-5: ]

array([ 86.98026156,  86.98026156,  99.6550156 ,  70.50593234,
       178.08445599])

In [ ]:
r2_score(y_test, preds)

-0.044013880736166966

In [ ]:
mean_absolute_error(y_test, preds)

468.14981196568084

In [ ]:
mean_squared_error(y_test, preds)

3725089.944541625

[Why is my r2 score negative?](https://stats.stackexchange.com/questions/183265/what-does-negative-r-squared-mean#:~:text=In%20actual%20practice%2C%20with%20ordinary,regression%20line%20must%20go%20through.&text=It%20just%20happens%20that%20that,gives%20a%20negative%20R%2DSquared.) Full post: https://stats.stackexchange.com/a/279358

The LinearRegression model was the best, but it had a low r2 score (0.39) and did not beat the baseline model. 

We should look into further expanding this dataset and collecting other parameters to help predict `price`.

Next idea: combine all 3 CSVs.